### ISSUE TITLE
cuse swtpm error on reading response TPM command on heavy load

### ISSUE THREAD

**Issue by marcos-antonio-moraes** [stefanberger/swtpm#725]
**Describe the bug**
Heavy load on cuse swtpm causes error on reading TPM command response phase.

**To Reproduce**
Loop

1 -  Create sapi context with tcti dev (/dev/tpm0)
2 - Get capabilities
3 -  Read target PCR
3 - Extend digest to target PCR
5 -  Read target PCR again to check previous extend operation
6 -  Release tcti/sapi context

**Expected behavior**
No error on executing the stress test.

**Desktop (please complete the following information):**
 - OS: Client written in C on Suse Enterprise SP3, using libtss2. Cuse swtpm built and running docker , based on Linux Alpine

**Versions of relevant components**
- swtpm:

**Log files**
Please attach any log files. If using a VM and it was started with libvirt, attach the logfile  found in `/var/log/swtpm/libvirt/qemu/VM-NAME-swtpm.log`.

**Additional context**
I am performing some stress testing with cuse swtpm and I found malfunction when stressing to much swtpm.  When swtpm prepares the response buffer it uses wrong offset. My finds point race condition caused by timing.

Debugging swtpm I found that the root cause is 2 'read' threads that are  executing function ptm_read_result that change ptm_read_offset variable.

I instrumented the code to help debugging as follow:

1st info is timestamp

2nd info is thead id

3rd,4th line number and file name. Do not rely on line number because my instrumentation changed line number form base code.

 

Notice that the issue shows up due to thread id 11

">> 2022-08-05 12:32:51.023492 - [11] [ 626:cuse_tpm.c]in ptm_read_result ptm_read_offset 19 len 19"
This execution increments the read offset that the other read trhead ( thread id 12) is about of using ptm_read_offset

---------

2022-08-05 12:32:51.023186 - [35] [ 974:cuse_tpm.c]ptm_req_len 53, ptm_res_len 0, ptm_res_tot 4096
2022-08-05 12:32:51.023207 - [35] [ 979:cuse_tpm.c]write in [53][800200000035000001820000000800000009400000090000000000000000010004000102030405060708090A0B0C0D0E0F10111213]
>>2022-08-05 12:32:51.023346 - [**11**] [ 596:cuse_tpm.c]ptm_read_offset 0,  ptm_res_len 19
>>2022-08-05 12:32:51.023366 - [**11**] [ 601:cuse_tpm.c]read out [19][80020000001300000000000000000000010000]
2022-08-05 12:32:51.023445 - [28] [ 974:cuse_tpm.c]ptm_req_len 20, ptm_res_len 0, ptm_res_tot 4096
2022-08-05 12:32:51.023464 - [28] [ 979:cuse_tpm.c]write in [20][8001000000140000017E00000001000403000100]
>> 2022-08-05 12:32:51.023492 - [**11**] [ 626:cuse_tpm.c]in ptm_read_result ptm_read_offset 19 len 19

 -------
The failure rate is very irregular: sometimes it fails at iteration 100, 500, 10000 and so on.

I changed a small piece of code and it solved the issue I was facing. Now the loop executes successfully: My last try was 500,000,000 ( five hundred million)  iterations.

**What I changed:**

 in cuse_tpm.c:

**From :**

```
 fuse_reply_buf(req, (const char *)&ptm_response[ ptm_read_offset   ], len);
 ptm_read_offset += len;
```

**To:**

```
+ size_t local_offset = ptm_read_offset;
+ ptm_read_offset += len;
+fuse_reply_buf(req, (const char *)&ptm_response[local_offset], len);
- fuse_reply_buf(req, (const char *)&ptm_response[ ptm_read_offset   ], len);
+// ptm_read_offset += len;
```

Basically, I incremented ptm_read_offset before calling fuse_reply_buf. This prevents the situation where 1st reading thread is pre-empted just after calling fuse_reply_buffer and it is resumed after some while executing the read offset increment after ‘writing’ thread had zeroed the reading offset.

For convenience I attached this client source code: tpm_stress.zip

[tpm_stress.zip](https://github.com/stefanberger/swtpm/files/9290466/tpm_stress.zip)

**Comment by stefanberger** [comment]
@marcos-antonio-moraes, thanks for posting the issue and the test tool.

The test tool needs to be built with this command:

```
g++ stress_tpm.c -o stress_tpm -I /usr/include/tss2/ -l tss2-sys -l tss2-tcti-device
```

I have now tried to recreate the issue before and after some modifications to the CUSE TPM that prevent multiple openers. I wasn't able to recreate the issue with 2.5 million iterations before and after these modifications that are now in master. Is `TPM_DEV_FILE=/dev/myvtpm0 ./stress_tpm 25000000` all one has to run? Can you try running it again with the latest master and tell me whether the issue still exists. It's quite possible that my machine is too slow to recreate the issue...

**Comment by stefanberger** [comment]
PR #725 should have the right solution for this problem now.

**Comment by marcos-antonio-moraes** [comment]
Hi Stefan,

I tried the branch master and the issue persists, although it seems slightly less susceptible to failures.

I forgot to mention that experiments that I made were base on latest releases of swtpm(v0.7.3) and libtpms (v0.9.5)

I’ve attached to this e-mail some data from the experiments with latest release, master, and master with the same fix that I applied in my earlier tests.


Thanks,

Marcos Antonio de Moraes

**Comment by marcos-antonio-moraes** [comment]
“Is TPM_DEV_FILE=/dev/myvtpm0 ./stress_tpm 25000000 all one has to run?”
Yes, it is.

From: Antonio de Moraes, Marcos
Sent: terça-feira, 9 de agosto de 2022 21:28
To: stefanberger/swtpm ***@***.***>
Subject: RE: [stefanberger/swtpm] cuse swtpm error on reading response TPM command on heavy load (Issue #725)

Hi Stefan,

I tried the branch master and the issue persists, although it seems slightly less susceptible to failures.

I forgot to mention that experiments that I made were base on latest releases of swtpm(v0.7.3) and libtpms (v0.9.5)

I’ve attached to this e-mail some data from the experiments with latest release, master, and master with the same fix that I applied in my earlier tests.


Thanks,

Marcos Antonio de Moraes

**Comment by stefanberger** [comment]
Any chance you can try it with this patch here as well. I think this is the right fix for this. Just the g_mutex_lock() and g_mutex_unlock() should be sufficient to apply.

```
From 26aaf89c800e5b15159523bca3f18b1da98c7014 Mon Sep 17 00:00:00 2001
From: Stefan Berger <stefanb@linux.ibm.com>
Date: Tue, 9 Aug 2022 19:58:33 -0400
Subject: [PATCH] swtpm: cuse: Extend usage of FILE_OPS_LOCK to protect a
 reading thread

Extend usage of the FILE_OPS_LOCK to prevent other threads from reading or
writing commands or doing ioctls while the current thread is reading a
response. This prevents a race condition where ptm_read_offset is set to 0
by a thread writing a new command to the device while the current thread
is reading a response from the device and needs this offset.

Resolves: https://github.com/stefanberger/swtpm/issues/725
Signed-off-by: Stefan Berger <stefanb@linux.ibm.com>
---
 src/swtpm/cuse_tpm.c | 9 +++++++--
 1 file changed, 7 insertions(+), 2 deletions(-)

diff --git a/src/swtpm/cuse_tpm.c b/src/swtpm/cuse_tpm.c
index 0ce992c7..e69500b2 100644
--- a/src/swtpm/cuse_tpm.c
+++ b/src/swtpm/cuse_tpm.c
@@ -545,6 +545,9 @@ static void ptm_read_result(fuse_req_t req, size_t size)
 {
     size_t len = 0;
 
+    /* prevent other threads from reading or writing cmds or doing ioctls */
+    g_mutex_lock(FILE_OPS_LOCK);
+
     if (tpm_running) {
         /* wait until results are ready */
         worker_thread_wait_done();
@@ -553,12 +556,14 @@ static void ptm_read_result(fuse_req_t req, size_t size)
     if (ptm_read_offset < ptm_res_len) {
         len = ptm_res_len - ptm_read_offset;
         if (size < len)
-           len = size;
+            len = size;
     }
 
     fuse_reply_buf(req, (const char *)&ptm_response[ptm_read_offset], len);
 
     ptm_read_offset += len;
+
+    g_mutex_unlock(FILE_OPS_LOCK);
 }
 
 /*
@@ -869,7 +874,7 @@ static void ptm_write_cmd(fuse_req_t req, const char *buf, size_t size,
     ptm_req_len = size;
     ptm_res_len = 0;
 
-    /* prevent other threads from writing or doing ioctls */
+    /* prevent other threads from reading or writing cmds or doing ioctls */
     g_mutex_lock(FILE_OPS_LOCK);
 
     if (tpm_running) {
```

**Comment by stefanberger** [comment]
I have now tried the test program again on a faster machine and it's not recreating the issue. Also , from what I can see this program doesn't use threads, so it's quite linear in what it's doing first write then read, so no writer will step on a read in this case...

**Comment by marcos-antonio-moraes** [comment]
Sure. I am going to try it right now!
